module/os/linux/spl/spl-zone.c

```diff
@@ -0,0 +1,324 @@
+	zds = zone_datasets_lookup(nsnum);
+	if (zds == NULL) {
+		zds = kmem_alloc(sizeof (zone_datasets_t), KM_SLEEP);
+		INIT_LIST_HEAD(&zds->zds_list);
+		INIT_LIST_HEAD(&zds->zds_datasets);
+		zds->zds_nsnum = nsnum;
+		list_add_tail(&zds->zds_list, &zone_datasets);
+	} else {
+		zd = zone_dataset_lookup(zds, dataset, dsnamelen);
+		if (zd != NULL) {
+			error = EEXIST;
+			goto done;
+		}
+	}
+
+	zd = kmem_alloc(sizeof (zone_dataset_t) + dsnamelen + 1, KM_SLEEP);
+	zd->zd_dsnamelen = dsnamelen;
+	strncpy(zd->zd_dsname, dataset, dsnamelen);
+	zd->zd_dsname[dsnamelen] = '\0';
+	INIT_LIST_HEAD(&zd->zd_list);
+	list_add_tail(&zd->zd_list, &zds->zds_datasets);
+	error = 0;
+
+done:
+	mutex_exit(&zone_datasets_lock);
+	return (error);
+#else
+	return (ENXIO);
+#endif
+}
+EXPORT_SYMBOL(zone_dataset_attach);
+
+int
+zone_dataset_detach(cred_t *cred, const char *dataset, unsigned int nsnum)
+{
+#if defined(CONFIG_USER_NS)
+	zone_datasets_t *zds;
+	zone_dataset_t *zd;
+	int error;
+	size_t dsnamelen;
+
+	if ((error = zone_dataset_cred_check(cred)) != 0)
+		return (error);
+	if ((error = zone_dataset_name_check(dataset, &dsnamelen)) != 0)
+		return (error);
+
+	mutex_enter(&zone_datasets_lock);
+	zds = zone_datasets_lookup(nsnum);
+	if (zds != NULL)
+		zd = zone_dataset_lookup(zds, dataset, dsnamelen);
+	if (zds == NULL || zd == NULL) {
+		error = ENOENT;
+		goto done;
+	}
+
+	list_del(&zd->zd_list);
+	kmem_free(zd, sizeof (*zd) + zd->zd_dsnamelen + 1);
+
+	/* Prune the namespace entry if it has no more delegations. */
+	if (list_empty(&zds->zds_datasets)) {
+		list_del(&zds->zds_list);
+		kmem_free(zds, sizeof (*zds));
+	}
+	error = 0;
+
+done:
+	mutex_exit(&zone_datasets_lock);
+	return (error);
+#else
+	return (ENXIO);
+#endif
+}
+EXPORT_SYMBOL(zone_dataset_detach);
+
+int
+zone_dataset_visible(const char *dataset, int *write)
+{
+	zone_datasets_t *zds;
+	zone_dataset_t *zd;
+	size_t dsnamelen, zd_len;
+
+	/* Default to read-only, in case visible is returned. */
+	if (write != NULL)
+		*write = 0;
+	if (zone_dataset_name_check(dataset, &dsnamelen) != 0)
+		return (0);
+	if (INGLOBALZONE(curproc)) {
+		if (write != NULL)
+			*write = 1;
+		return (1);
+	}
+
+	mutex_enter(&zone_datasets_lock);
+	int visible = 0;
+	zds = zone_datasets_lookup(crgetzoneid(curproc->cred));
+	if (zds == NULL)
+		goto done;
+
+	list_for_each_entry(zd, &zds->zds_datasets, zd_list) {
+		zd_len = strlen(zd->zd_dsname);
+		/*
+		 * A dataset is writable only if it matches one of the
+		 * namespace's entries, or is a child of an entry.
+		 * The parent datasets of namespace entries are
+		 * visible, but read-only, to provide a path back to the
+		 * root of the pool.
+		 */
+		if (zd_len > dsnamelen) {
+			visible = memcmp(zd->zd_dsname, dataset,
+			    dsnamelen) == 0 &&
+			    zd->zd_dsname[dsnamelen] == '/';
+		} else {
+			visible = memcmp(zd->zd_dsname, dataset, zd_len) == 0;
+			if (visible) {
+				if (write != NULL)
+					*write = 1;
+			}
+		}
+	}
```

Comment:
This loop seems to be missing a break statement otherwise later iterations of this loop will reset the visibility of datasets back to invisible.
I delegated two datasets that have distinctive names and could only see the one that was delegated last inside the namespace.

Also, datasets, where the name has the same prefix as a delegated one, are incorrectly considered visible.

```
[root@zvm-arch ~]# zfs list 
NAME          USED AVAIL REFER MOUNTPOINT 
pool         1.67M 18.9G  112K /pool 
pool/data      96K 18.9G   96K /pool/data 
pool/data2     96K 18.9G   96K /pool/data2 
pool/dataset   96K 18.9G   96K /pool/dataset 
pool/new       96K 18.9G   96K /pool/new 
[root@zvm-arch ~]# unshare -Urm sh -c "sleep infinity" & 
[root@zvm-arch ~]# nsenter --preserve-credentials --user --mount -t $NS_PID zfs list 
no datasets available 
[root@zvm-arch ~]# zfs zone $NS pool/data
[root@zvm-arch ~]# nsenter --preserve-credentials --user --mount -t $NS_PID zfs list 
NAME          USED AVAIL REFER MOUNTPOINT 
pool         1.67M 18.9G  112K /pool 
pool/data      96K 18.9G   96K /pool/data 
pool/data2     96K 18.9G   96K /pool/data2 
pool/dataset   96K 18.9G   96K /pool/dataset
 ```

pool/dataset and pool/data2 are incorrectly visible inside the namespace.

Reply:
It seems this was caused by a previous change to the if ladder to try to silence a compiler warning and had an undesired side effect. Should be fixed in the new version